File: OneSignalSDK/onesignal/src/main/java/com/onesignal/influence/OSTrackerFactory.java
package com.onesignal.influence;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.onesignal.OSLogger;
import com.onesignal.OSSharedPreferences;
import com.onesignal.OneSignal;
import com.onesignal.OneSignalRemoteParams;
import com.onesignal.influence.model.OSInfluence;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class OSTrackerFactory {

    private ConcurrentHashMap<String, OSChannelTracker> trackers = new ConcurrentHashMap<>();

    private OSInfluenceDataRepository dataRepository;

    public OSTrackerFactory(OSSharedPreferences preferences, OSLogger logger) {
        dataRepository = new OSInfluenceDataRepository(preferences);

        trackers.put(OSInAppMessageTracker.TAG, new OSInAppMessageTracker(dataRepository, logger));
        trackers.put(OSNotificationTracker.TAG, new OSNotificationTracker(dataRepository, logger));
    }

    public void saveInfluenceParams(OneSignalRemoteParams.InfluenceParams influenceParams) {
        dataRepository.saveInfluenceParams(influenceParams);
    }

    public void addSessionData(@NonNull JSONObject jsonObject, List<OSInfluence> influences) {
        for (OSInfluence influence : influences) {
            switch (influence.getInfluenceChannel()) {
                case NOTIFICATION:
                    getNotificationChannelTracker().addSessionData(jsonObject, influence);
                case IAM:
                    // We don't track IAM session data for on_focus call
            }
        }
    }

    public void initFromCache() {
        for (OSChannelTracker tracker : trackers.values()) {
            tracker.initInfluencedTypeFromCache();
        }
    }

    public List<OSInfluence> getInfluences() {
        List<OSInfluence> influences = new ArrayList<>();
        for (OSChannelTracker tracker : trackers.values()) {
            influences.add(tracker.getCurrentSessionInfluence());
        }
        return influences;
    }

    public List<OSInfluence> getSessionInfluences() {
        List<OSInfluence> influences = new ArrayList<>();
        for (OSChannelTracker tracker : trackers.values()) {
            // IAM doesn't influence session calls
            if (tracker instanceof OSInAppMessageTracker)
                continue;
            influences.add(tracker.getCurrentSessionInfluence());
        }
        return influences;
    }

    public OSChannelTracker getIAMChannelTracker() {
        return trackers.get(OSInAppMessageTracker.TAG);
    }

    public OSChannelTracker getNotificationChannelTracker() {
        return trackers.get(OSNotificationTracker.TAG);
    }

    @Nullable
    public OSChannelTracker getChannelByEntryAction(OneSignal.AppEntryAction entryAction) {
        if (entryAction.isNotificationClick())
            return getNotificationChannelTracker();

        return null;
    }

    public List<OSChannelTracker> getChannels() {
        List<OSChannelTracker> channels = new ArrayList<>();

        OSChannelTracker notificationChannel = getNotificationChannelTracker();
        if (notificationChannel != null)
            channels.add(notificationChannel);

        OSChannelTracker iamChannel = getIAMChannelTracker();
        if (iamChannel != null)
            channels.add(iamChannel);

        return channels;
    }

    public List<OSChannelTracker> getChannelsToResetByEntryAction(OneSignal.AppEntryAction entryAction) {
        List<OSChannelTracker> channels = new ArrayList<>();

        // Avoid reset session if application is closed
        if (entryAction.isAppClose())
            return channels;

        // Avoid reset session if app was focused due to a notification click (direct session recently set)
        if (entryAction.isAppOpen()) {
            OSChannelTracker notificationChannel = getNotificationChannelTracker();
            if (notificationChannel != null)
                channels.add(notificationChannel);
        }

        OSChannelTracker iamChannel = getIAMChannelTracker();
        if (iamChannel != null)
            channels.add(iamChannel);

        return channels;
    }

}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/influence/data/OSTrackerFactory.kt
package com.onesignal.influence.data

import com.onesignal.OSLogger
import com.onesignal.OSSharedPreferences
import com.onesignal.OSTime
import com.onesignal.OneSignal.AppEntryAction
import com.onesignal.OneSignalRemoteParams.InfluenceParams
import com.onesignal.influence.OSInfluenceConstants
import com.onesignal.influence.domain.OSInfluence
import com.onesignal.influence.domain.OSInfluenceChannel
import org.json.JSONObject
import java.util.*
import java.util.concurrent.ConcurrentHashMap

class OSTrackerFactory(preferences: OSSharedPreferences, logger: OSLogger, timeProvider: OSTime) {
    private val trackers = ConcurrentHashMap<String, OSChannelTracker>()
    private val dataRepository: OSInfluenceDataRepository = OSInfluenceDataRepository(preferences)

    val influences: List<OSInfluence>
        get() = trackers.values.map { it.currentSessionInfluence }

    val sessionInfluences: List<OSInfluence>
        get() = trackers.values.filter { it.idTag != OSInfluenceConstants.IAM_TAG }.map { it.currentSessionInfluence }

    val iAMChannelTracker: OSChannelTracker
        get() = trackers[OSInfluenceConstants.IAM_TAG]!!

    val notificationChannelTracker: OSChannelTracker
        get() = trackers[OSInfluenceConstants.NOTIFICATION_TAG]!!

    val channels: List<OSChannelTracker>
        get() {
            val channels: MutableList<OSChannelTracker> = mutableListOf()
            notificationChannelTracker.let { channels.add(it) }
            iAMChannelTracker.let { channels.add(it) }
            return channels
        }

    init {
        trackers[OSInfluenceConstants.IAM_TAG] = OSInAppMessageTracker(dataRepository, logger, timeProvider)
        trackers[OSInfluenceConstants.NOTIFICATION_TAG] = OSNotificationTracker(dataRepository, logger, timeProvider)
    }

    fun initFromCache() {
        trackers.values.forEach {
            it.initInfluencedTypeFromCache()
        }
    }

    fun saveInfluenceParams(influenceParams: InfluenceParams) {
        dataRepository.saveInfluenceParams(influenceParams)
    }

    fun addSessionData(jsonObject: JSONObject, influences: List<OSInfluence>) {
        influences.forEach {
            when (it.influenceChannel) {
                OSInfluenceChannel.NOTIFICATION -> notificationChannelTracker.addSessionData(jsonObject, it)
                OSInfluenceChannel.IAM -> {
                    // IAM doesn't influence session
                }
            }
        }
    }

    fun getChannelByEntryAction(entryAction: AppEntryAction): OSChannelTracker? {
        return if (entryAction.isNotificationClick) notificationChannelTracker else null
    }

    fun getChannelsToResetByEntryAction(entryAction: AppEntryAction): List<OSChannelTracker> {
        val channels: MutableList<OSChannelTracker> = ArrayList()
        // Avoid reset session if application is closed
        if (entryAction.isAppClose) return channels
        // Avoid reset session if app was focused due to a notification click (direct session recently set)
        val notificationChannel = if (entryAction.isAppOpen) notificationChannelTracker else null
        notificationChannel?.let {
            channels.add(it)
        }
        iAMChannelTracker.let {
            channels.add(it)
        }
        return channels
    }
}